src/window.rs

```diff
@@ -1,73 +1,93 @@
+use {Float, Sample};
+use core;
 use core::marker::PhantomData;
-use core::f64;
-
-use {Sample, FloatSample};
-use signal;
-use signal::{Phase, ConstHz};
-use conv::{FromSample, ToSample};
 use frame::Frame;
+use signal::{self, Signal};
+
 
-// Using types instead of enum allows for static dispatch
-pub trait Type<S: Sample> {
-    fn at_phase(phase: S) -> S;
+/// The window function used within a `Window`.
+pub trait Type {
+    /// Returns the amplitude for the given phase, given as some `Sample` type.
+    fn at_phase<S: Sample>(phase: S) -> S;
 }
 
+
+/// A type of window function, also known as teh "raised cosine window".
+///
+/// [Wiki entry](https://en.wikipedia.org/wiki/Window_function#Hann_.28Hanning.29_window).
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
 pub struct Hanning;
 
+/// The simplest window type, equivalent to replacing all but *N* values of data sequence by
+/// zeroes, making it appear as though the waveform suddenly turns on and off.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
 pub struct Rectangle;
 
-pub struct Window<F, WT> 
+/// A `Signal` type that for every yielded `phase`, yields the amplitude across the `window::Type`
+/// for that phase.
+#[derive(Clone)]
+pub struct Window<F, W> 
     where F: Frame,
-          WT: Type<F::Sample>
+          W: Type,
 {
-    pub phase: Phase<ConstHz>,
+    /// Yields phase stepped at a constant rate to be passed to the window function `W`.
+    pub phase: signal::Phase<signal::ConstHz>,
     ftype: PhantomData<F>,
-    wttype: PhantomData<WT>
+    wttype: PhantomData<W>
 }
 
-pub struct WindowedFrame<'a, F, WT> 
-    where F: 'a + Frame,
-          F::Sample: FromSample<f64>, 
-          WT: Type<F::Sample>
+/// An iterator that multiplies frames yielded by some `Signal` with a `Window`.
+///
+/// Yields `Windowed` frames.
+#[derive(Clone)]
+pub struct Windowed<S, W> 
+    where S: Signal,
+          S::Item: Frame,
+          W: Type,
 {
-    pub data: &'a [F],
-    pub window: Window<F, WT>,
-    idx: usize
+    signal_with_window: core::iter::Zip<S, Window<S::Item, W>>,
 }
 
-/// Windower takes a long slice of data and generates an iterator over its frames
-pub struct Windower<'a, F, WT> 
+/// Takes a long slice of frames and yields `Windowed` chunks of size `bin` once every `hop` frames.
+#[derive(Clone)]
+pub struct Windower<'a, F, W> 
     where F: 'a + Frame, 
-          F::Sample: FromSample<f64>, 
-          WT: Type<F::Sample>
+          W: Type,
 {
+    /// The size of each `Windowed` chunk to be yielded.
     pub bin: usize,
+    /// The step size over `frames` for the start of each new `Windowed` chunk.
     pub hop: usize,
-    pub idx: usize,
```

Comment:
Hmmm, my thoughts were that making them private might ensure a little more invariance in the behaviour of the `Windower` (I actually meant to make `bin` and `hop` private too), and that it's almost as easy/efficient to construct a new `Windower` as it is to mutate the fields anyway. Now that you mention it though, I can't see much harm being done in having them public, and it seems like you might have  a particular use-case in mind so I'm happy to re-add them!